Implement copy_, fill_, and ones_like for Nested Tensors backends

ani300 opened the pull request:
Summary: This diff implements copy_ in order to allow pinned memory transfers for nested tensors, as well as fill_ and ones_like, to test whether nested tensors can be created with other factory functions.

Test Plan: Pass all CI and sandcastle jobs.

Reviewed By: mikekgfb

Differential Revision: D40689594

cpuhrsch commented on torch/nested/__init__.py:
```diff
@@ -41,7 +41,7 @@ def nested_tensor(tensor_list: List[Tensor], *, dtype: Optional[DType] = None, d
         True
     """
-    if not isinstance(tensor_list, list) or any([not torch.is_tensor(t) for t in tensor_list]):
+    if not isinstance(tensor_list, list) or any([not isinstance(t, torch.Tensor) for t in tensor_list]):
```
Why this change?

ani300 replied:
I think it was done automatically from a linter suggestion, I can revert it

cpuhrsch commented on test/test_nestedtensor.py:
```diff
@@ -1,18 +1,26 @@
 # Owner(s): ["module: nestedtensor"]
 
+import unittest
```
Can we add unit tests for these ops?

ani300 replied:
I'll add them

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorFactories.cpp:
```diff
@@ -106,9 +106,26 @@ Tensor _to_copy_nested(
+
+Tensor& fill_nested_(Tensor& self, const Tensor& value) {
+  const auto& self_buf = get_nested_tensor_impl(self)->get_buffer();
+  self_buf.fill_(value);
```
This won't work if value is anything but a scalar. In the test you only all it with scalar. Do you need to support this overload for your use case?

ani300 replied:
No. I can add the test that fails with a non-0-dimensional tensor too. I think the reason this exists is so that you can do something like t.fill_(other_t[0]), which returns a 0-dimensional tensor instead of a scalar.

Commit:
Implement copy_, fill_, and ones_like for Nested Tensors backends (#87728)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/87728

This diff implements copy_ in order to allow pinned memory transfers for nested tensors, as well as fill_ and ones_like, to test whether nested tensors can be created with other factory functions.

Test Plan: Pass all CI and sandcastle jobs.

Reviewed By: mikekgfb

Differential Revision: D40689594

fbshipit-source-id: f1e180b40d1c7317768da7bc8f8abd68ab26dd6b

ani300 commented:
@pytorchbot merge

ani300 commented:
@pytorchbot merge -f "unrelated failure"
